### PR TITLE
Support gradual increase learning rate in learningratescheduler

### PR DESCRIPTION
**Pull request by dding3**
## What changes were proposed in this pull request?
1. Support gradual increase learning rate for learningratescheduler
2. Support user set gradual increase lr in Inception example

## How was this patch tested?
manual test

### CODE REVIEW

**Comment by jason-dai** [comment]
Instead make the change specific to Poly and EpochDecay, we should make it a general scheme so that this can be applied to different LearningRateSchedule

File: spark/dl/src/main/scala/com/intel/analytics/bigdl/models/inception/Options.scala [hunk 1]
```diff
@@ -79,6 +84,21 @@ object Options {
     opt[Unit]('g', "graphModel")
       .text("use graph model")
       .action((x, c) => c.copy(graphModel = true))
+    opt[Double]("maxLr")
```
Comment (qiuxin2012) [inline spark/dl/src/main/scala/com/intel/analytics/bigdl/models/inception/Options.scala]: update readme for these new params?

File: spark/dl/src/main/scala/com/intel/analytics/bigdl/optim/SGD.scala [hunk 2]
```diff
@@ -211,6 +213,10 @@ object SGD {
     def updateHyperParameter(config : Table, state : Table) : Unit = {}
 
     var currentRate : Double = 0.0
+
+    var minusIteration : Int = 0
```
Comment (qiuxin2012) [inline spark/dl/src/main/scala/com/intel/analytics/bigdl/optim/SGD.scala]: Add some comment?

File: spark/dl/src/main/scala/com/intel/analytics/bigdl/optim/SGD.scala [hunk 3]
```diff
@@ -579,4 +595,83 @@ object SGD {
...
+
+    override def updateHyperParameter(config: Table, state: Table): Unit = {
+      val nevals = state.get[Int]("evalCounter").getOrElse(0)
+
+      if (cur == 0 && startLr == 0) {
+        startLr = config.get[Double]("learningRate").getOrElse(1e-3)
+        schedules(cur).startLr = startLr
+      }
+
+      while (nevals > iterations(cur)) {
```
Comment (qiuxin2012) [inline spark/dl/src/main/scala/com/intel/analytics/bigdl/optim/SGD.scala]: Please merge the duplicate code

File: spark/dl/src/main/scala/com/intel/analytics/bigdl/optim/SGD.scala [hunk 4]
```diff
@@ -211,6 +213,12 @@ object SGD {
     def updateHyperParameter(config : Table, state : Table) : Unit = {}
 
     var currentRate : Double = 0.0
+
+    // iteration numbers needed to be excluded for a new learningRateSchedule
+    var excludeIterations : Int = 0
+    // epoch numbers needed to be excluded for a new learningRateSchedule
+    var excludeEpochs: Int = 0
+    var startLr: Double = 0.0
```
Comment (qiuxin2012) [inline spark/dl/src/main/scala/com/intel/analytics/bigdl/optim/SGD.scala]: protected?

File: spark/dl/src/main/scala/com/intel/analytics/bigdl/optim/SGD.scala [hunk 5]
```diff
@@ -579,4 +597,78 @@ object SGD {
...
+      val nevals = optimMethod.state.get[Int]("evalCounter").getOrElse(0)
+      val lr = if (startLr == 0) optimMethod.learningRate else startLr
+      val clr = - lr - delta * (nevals - excludeIterations)
+      currentRate = clr
+      println(s"iteration is : ${nevals}. current learning rate is $clr")
+      optimMethod.state("evalCounter") = nevals + 1
+    }
+  }
+
+  case class SequentialSchedule(iterationPerEpoch: Int) extends LearningRateSchedule {
```
Comment (qiuxin2012) [inline spark/dl/src/main/scala/com/intel/analytics/bigdl/optim/SGD.scala]: doc

File: spark/dl/src/main/scala/com/intel/analytics/bigdl/optim/SGD.scala [hunk 6]
```diff
@@ -579,4 +597,78 @@ object SGD {
...
+    val schedules: ArrayBuffer[LearningRateSchedule] = ArrayBuffer[LearningRateSchedule]()
+    val iterations: ArrayBuffer[Int] = ArrayBuffer[Int]()
+    var cur: Int = 0
+
+    def add(schedule: LearningRateSchedule, maxIteration: Int = Int.MaxValue):
```
Comment (qiuxin2012) [inline spark/dl/src/main/scala/com/intel/analytics/bigdl/optim/SGD.scala]: all public method should have a doc

File: pyspark/bigdl/optim/optimizer.py [hunk 7]
```diff
@@ -316,6 +315,31 @@ def __init__(self,
...
+
+    :param delta: increase amount after each iteration
+
+    >>> warmup = Warmup(0.05)
+    creating: createWarmup
+    """
+    def __init__(self, delta, bigdl_type="float"):
+        JavaValue.__init__(self, None, bigdl_type, delta)
+
+class SequentialSchedule(JavaValue):
```
Comment (qiuxin2012) [inline pyspark/bigdl/optim/optimizer.py]: `add` function?

File: spark/dl/src/main/scala/com/intel/analytics/bigdl/models/inception/Options.scala (hunk 1 above)
Comment (dding3) [reply spark/dl/src/main/scala/com/intel/analytics/bigdl/models/inception/Options.scala]: Updated

File: pyspark/bigdl/optim/optimizer.py (hunk 7 above)
Comment (dding3) [reply pyspark/bigdl/optim/optimizer.py]: Added

File: pyspark/bigdl/optim/optimizer.py [hunk 8]
```diff
@@ -316,6 +315,41 @@ def __init__(self,
...
+
+    :param iterationPerEpoch: iteration numbers per epoch
+
+    >>> sequentialSchedule = SequentialSchedule(5)
+    creating: SequentialSchedule
+    """
+    def __init__(self, iteration_per_epoch, bigdl_type="float"):
+        JavaValue.__init__(self, None, bigdl_type, iteration_per_epoch)
+
+    def add(self, scheduler, max_iteration=sys.maxsize, bigdl_type="float"):
```
Comment (qiuxin2012) [inline pyspark/bigdl/optim/optimizer.py]: add test for this `add`

File: pyspark/bigdl/optim/optimizer.py [hunk 9]
```diff
@@ -356,6 +390,9 @@ class SGD(OptimMethod):
     :param nesterov enables Nesterov momentum
     :param learningrates 1D tensor of individual learning rates
     :param weightdecays 1D tensor of individual weight decays
+    :param warmup_iteration iteration numbers to take for learning rate reach to max learning rate
```
Comment (qiuxin2012) [inline pyspark/bigdl/optim/optimizer.py]: Couldn't find corresponding parameters in SGD.

File: spark/dl/src/main/scala/com/intel/analytics/bigdl/optim/SGD.scala [hunk 10]
```diff
@@ -579,4 +589,76 @@ object SGD {
...
+    var cur: Int = 0
+
+    /**
+     * Add a learning rate scheduler to the contained `schedules`
+     *
+     * @param schedule learning rate scheduler to be add
+     * @param maxIteration iteration numbers this scheduler will run
+     * @return this container
+     */
+    def add(schedule: LearningRateSchedule, maxIteration: Int = Int.MaxValue):
```
Comment (qiuxin2012) [inline spark/dl/src/main/scala/com/intel/analytics/bigdl/optim/SGD.scala]: Why set maxIteration to Int.MaxValue?

**Comment by qiuxin2012** [comment]
also need to update https://github.com/intel-analytics/BigDL/blob/master/docs/docs/APIGuide/Optimizers/Optim-Methods.md

**Comment by dding3** [comment]
PR validation pass. Thank you for your review!

**Comment by jason-dai** [comment]
We should add doc for learning rate schdule

**Pull request merged**